refactor(storage): separate each sample into a separate file

laljikanjareeya opened the pull request:
Fixes #930

frankyn commented:
@JesseLovelace is lead on sample split work, and is working on adding samples to documentation. It will take time to complete and they'll update the PR when ready to merge.

We can't merge until that's ready.

JesseLovelace commented on storage/api/Storage/ListFiles.cs:
```diff
@@ -0,0 +1,53 @@
+        }
+        // [END storage_list_files]
+
+        // [START storage_list_files_with_prefix]
```
This should be a separate class, and should have documentation explaining how listing with delimiters works. See the other language versions of this sample for examples on what that documentation can look like. i.e. https://cloud.google.com/storage/docs/listing-objects#storage-list-objects-go

laljikanjareeya replied:
Done!!

JesseLovelace commented on storage/api/Storage/ListFilesWithPrefix.cs:
```diff
@@ -0,0 +1,56 @@
+        //   /a/1.txt
+        //   /a/b/2.txt
+        //
+        // However, if you specify prefix="a/" and delim="/", you'll get back:
```
"delim" should be "delimiter" here

laljikanjareeya replied:
Done!

laljikanjareeya commented:
> Please make sure that _each_ sample prints something. For example, DisableRequesterPays should print something like "Requester pays disabled for {bucketName}". There's a few samples here missing print statements, feel free to look at the other language samples for example of what to print.

Added Prints where it was missing plz take another look.

laljikanjareeya commented:
@JesseLovelace Resolved all the comments and added Print in all the sample Plz take as look.

laljikanjareeya commented:
@JesseLovelace Addressed all the comments could you please take a look.

laljikanjareeya commented:
@frankyn can you take a look.

frankyn commented:
Thank you for your patience @laljikanjareeya, @JesseLovelace will be reviewing this change based on today's meeting.

amanda-tarafa commented:
I'll re-review as well, given that our canonical was improved after the PR was created. I think it's almost there though so shouldn't be a problem. I'll try to get to it tomorrow.

laljikanjareeya commented:
> I'll re-review as well, given that our canonical was improved after the PR was created. I think it's almost there though so shouldn't be a problem. I'll try to get to it tomorrow.

@amanda-tarafa  I see that this PR is too old so it's missing our new canonical as we did in PubSub Samples, It also have command line runner.
Wait for a while I will finish it soon and than you can start review.

laljikanjareeya commented:
@amanda-tarafa @JesseLovelace  Now it's ready for review.
Three test cases are failing in that two is due to project is exceeding the rate limit for creating and deleting buckets and one is because of KMS region Please take a look.
Thanks!!

JesseLovelace commented on storage/api/Storage.Samples/RemoveBucketIamMember.cs:
```diff
@@ -0,0 +1,47 @@
+
+public class RemoveBucketIamMemberSample
+{
+    public void RemoveBucketIamMember(string bucketName, string role, string member)
```
See the comment I left in the add iam member sample about the role and member variables

JesseLovelace commented on storage/api/Storage.Samples/MakePublic.cs:
```diff
@@ -0,0 +1,38 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START storage_make_public]
```
For our consistency pass on region tags, please replace storage_make_public with storage_set_bucket_public_iam

JesseLovelace commented on storage/api/Storage.Samples/GenerateEncryptionKey.cs:
```diff
@@ -0,0 +1,27 @@
+    public string GenerateEncryptionKey()
+    {
+        var encryptionKey = EncryptionKey.Generate().Base64Key;
+        return encryptionKey;
```
Print out something like "Generated Base64-encoded AES-256 encryption key: {encryptionKey}"

JesseLovelace commented on storage/api/Storage.Samples/RemoveBucketIamMember.cs:
```diff
@@ -0,0 +1,47 @@
+        // Set the modified IAM policy to be the current IAM policy.
+        storage.SetBucketIamPolicy(bucketName, policy);
+        Console.WriteLine($"Removed {member} with role {role} "
+            + $"to {bucketName}");
```
should be "from" instead of  "to"

amanda-tarafa commented:
@JesseLovelace as per your comment:

> 1. For every argument in a sample, there needs to be a comment giving an example of what that argument could be. For example, if "bucketName" is the first argument in a sample, that sample should start with a comment like `// string bucketName = "your-bucket-name";`, if an encryption key is an argument there should be a comment like `// string key = "TIbv/fjexq+VmtXzAlc63J4z5kFmWJ6NdAPQulQBT7g=";`

We usually don't add that comment on the sample code because if the user were to copy the sample (which is what they mostly do) and uncomment those lines, the sample wouldn't build because you already have parameters named the same. And we want the user experience to be as straightforward as possible.
But what we can do is put in those sample values as default parameter values when it's possible. If it's not possible (because of the parameter type), we can add XML documentation to the sample explaining what the parameter is:

Default value for parameters would look like:

`public void SomeSample(string projectId = "my-project-id", string bucket-name = "my-bucket-name")`

XML doc would look like

```
    /// <summary>
    /// This sample shows something
    /// </summary>
    /// <param name="projectId">Explanation of param</param>
    /// <param nanme="bucketList">Explanation of param that cannot have a default value</param>
    public void SomeSample(string projectId, IList<string> bucketList)
```

Hope this is OK with you since either option meets the objective of exemplifying the parameter.
Thanks.

amanda-tarafa commented:
@JesseLovelace as per your second general comment:

`Every sample should have projectId as a variable, and should set up the storage client to be associated with that project ID. For example, every Java sample has projectId as a variable and initiates the client as Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();`

There's no way to create a Storage client passing in a project ID in the .NET library. You need to pass the project ID on each operation, either implicitly (when passing a bucket resource name for instance) or explicitly.

amanda-tarafa commented:
For the failures due to bucket creation/deletion being rate limited, see [here](https://github.com/googleapis/google-cloud-dotnet/blob/8ecc511a287160a52ee3e7d4abd04862696fb8d3/apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/StorageFixture.cs#L271) what we have on the integration tests in client libraries. And we call that method whenever we create or delete a bucket. You can do something similar.

JesseLovelace commented on storage/api/Storage.Samples/AddBucketIamMember.cs:
```diff
@@ -0,0 +1,40 @@
+
+public class AddBucketIamMemberSample
+{
+    public Policy AddBucketIamMember(string bucketName, string role, string member)
```
Yes, this makes sense to me. I'd still add the link to the documentation, but the important thing is showing an example of what values are acceptable so this works for that

JesseLovelace replied:
Yes, I'm okay with that

JesseLovelace commented:
> @JesseLovelace as per your comment:
> 
> > 1. For every argument in a sample, there needs to be a comment giving an example of what that argument could be. For example, if "bucketName" is the first argument in a sample, that sample should start with a comment like `// string bucketName = "your-bucket-name";`, if an encryption key is an argument there should be a comment like `// string key = "TIbv/fjexq+VmtXzAlc63J4z5kFmWJ6NdAPQulQBT7g=";`
> 
> We usually don't add that comment on the sample code because if the user were to copy the sample (which is what they mostly do) and uncomment those lines, the sample wouldn't build because you already have parameters named the same. And we want the user experience to be as straightforward as possible.
> But what we can do is put in those sample values as default parameter values when it's possible. If it's not possible (because of the parameter type), we can add XML documentation to the sample explaining what the parameter is:
> 
> Default value for parameters would look like:
> 
> `public void SomeSample(string projectId = "my-project-id", string bucket-name = "my-bucket-name")`
> 
> XML doc would look like
> 
> ```
>     /// <summary>
>     /// This sample shows something
>     /// </summary>
>     /// <param name="projectId">Explanation of param</param>
>     /// <param nanme="bucketList">Explanation of param that cannot have a default value</param>
>     public void SomeSample(string projectId, IList<string> bucketList)
> ```
> 
> Hope this is OK with you since either option meets the objective of exemplifying the parameter.
> Thanks.

Generally this is okay with me, but I don't just want an explanation of what the parameter can be, I want an example. Would it work to include commented code that reassigns the variable instead of declaring a new one with the same name?



> @JesseLovelace as per your second general comment:
> 
> > Every sample should have projectId as a variable, and should set up the storage client to be associated with that project ID. For example, every Java sample has projectId as a variable and initiates the client as Storage storage = StorageOptions.newBuilder().setProjectId(projectId).build().getService();
> 
> There's no way to create a Storage client passing in a project ID in the .NET library. You need to pass the project ID on each operation, either implicitly (when passing a bucket resource name for instance) or explicitly.

Thanks for letting me know that, in that case just make sure to include it where it makes sense.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/BucketIamMemberTest.cs:
```diff
@@ -0,0 +1,49 @@
+using Xunit;
+
+[Collection(nameof(BucketFixture))]
+public class BucketIamMemberTest
```
Testing three functionality here 
1. AddBucketIamMembers
2. RemoveBucketIamMembers
3. ViewBucketIamMembers

laljikanjareeya replied:
Separate the tests

laljikanjareeya commented on storage/api/Storage.Samples/ViewBucketIamMembers.cs:
```diff
@@ -0,0 +1,39 @@
+
+public class ViewBucketIamMembersSample
+{
+    public Policy ViewBucketIamMembers(string bucketName)
```
It has only bucket name as string and `GetBucketIamPolicyOptions`.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/BucketIamMemberTest.cs:
```diff
@@ -0,0 +1,49 @@
+        // remove bucket Iam member
+        removeBucketIamMemberSample.RemoveBucketIamMember(_bucketFixture.BucketName, role, $"{memberType}:{_bucketFixture.ServiceAccountEmail}");
+
+        result = viewBucketIamMembersSample.ViewBucketIamMembers(_bucketFixture.BucketName);
+        Assert.DoesNotContain(result.Bindings, cs => cs.Role == role);
+        Assert.DoesNotContain(result.Bindings.SelectMany(c => c.Members), c => c == $"{memberType}:{_bucketFixture.ServiceAccountEmail}");
```
I am testing all the three functionality here as it's related.

laljikanjareeya replied:
separate the tests

JesseLovelace commented on storage/api/Storage.Samples/ListFilesWithPrefix.cs:
```diff
@@ -0,0 +1,61 @@
+    ///   /a/1.txt
+    ///   /a/b/2.txt
+    ///
+    /// However, if you specify prefix="a/" and delimiter="/", you'll get back:
```
Please show what they would get back, looks like it got cut off here

laljikanjareeya replied:
Done Thanks!

JesseLovelace commented on storage/api/Storage.Samples/DisableUniformBucketLevelAccess.cs:
```diff
@@ -0,0 +1,45 @@
+        var storage = StorageClient.Create();
+        var bucket = storage.GetBucket(bucketName);
+        bucket.IamConfiguration.UniformBucketLevelAccess.Enabled = false;
+        /** THIS IS A WORKAROUND */
```
I don't think we need this workaround anymore.

amanda-tarafa commented on storage/api/Storage.Samples.Tests/CopyFileTest.cs:
```diff
@@ -0,0 +1,37 @@
+    {
+        CopyFileSample copyFileSample = new CopyFileSample();
+        ListFilesSample listFilesSample = new ListFilesSample();
+        copyFileSample.CopyFile(_bucketFixture.BucketNameGeneric, _bucketFixture.FileName, _bucketFixture.BucketNameRegional,
+            _bucketFixture.CollectRegionalObject("CopyFile.txt"));
```
```suggestion
        copyFileSample.CopyFile(
            _bucketFixture.BucketNameGeneric, _bucketFixture.FileName, _bucketFixture.BucketNameRegional, _bucketFixture.CollectRegionalObject("CopyFile.txt"));
```

amanda-tarafa commented on storage/api/Storage.Samples.Tests/BucketFixture.cs:
```diff
@@ -0,0 +1,155 @@
+        uploadFileSample.UploadFile(BucketNameGeneric, FilePath);
+
+        Collect(FileName);
+    }
+    public void Dispose()
```
```suggestion
    }
    
    public void Dispose()
```

amanda-tarafa commented on storage/api/Storage.Samples.Tests/BucketFixture.cs:
```diff
@@ -0,0 +1,155 @@
+        }
+    }
+
+    internal void DeleteAllHmacKeys()
```
I don't know where you are calling this. But remember that tests run roughtly at the same time in Windows and Linux, and this method running in one environment could definetely interfere with the other environment.

laljikanjareeya replied:
Removed!!

JesseLovelace commented:
+1 to removing the XML doc that just adds noise. Keep the ones that aren't immediately intuitive or that add additional information, but no need to basically say the same thing twice

laljikanjareeya commented:
> I've reviewed some of the samples and their corresponding tests. I'll continue tomorrow. Mostly fine.

@amanda-tarafa Thanks for the review, Let me know once you finish the review I will make review changes once you finish the review, I will keep the XML docs where it adds significant information.

amanda-tarafa commented:
> > I've reviewed some of the samples and their corresponding tests. I'll continue tomorrow. Mostly fine.
> 
> @amanda-tarafa Thanks for the review, Let me know once you finish the review I will make review changes once you finish the review, I will keep the XML docs where it adds significant information.

Yes, will let you know, thanks!

amanda-tarafa commented:
Just and update, I'm halfway there and will finish the review on Monday.

amanda-tarafa commented on storage/api/Storage.Samples/DownloadByteRange.cs:
```diff
@@ -0,0 +1,60 @@
+        string localPath = null)
+    {
+        var storageClient = StorageClient.Create();
+        localPath = localPath ?? $"{Path.GetFileName(objectName)}_{firstByte}-{lastByte}";
```
Maybe move this code to the test and assume you are receiving a valid local path.

amanda-tarafa commented on storage/api/Storage.Samples/DownloadByteRange.cs:
```diff
@@ -0,0 +1,60 @@
+        string objectName = "my-file-name",
+        long firstByte = 0,
+        long lastByte = 20,
+        string localPath = null)
```
```suggestion
        string localPath = "local-file-path")
```

amanda-tarafa commented on storage/api/Storage.Samples/DownloadByteRange.cs:
```diff
@@ -0,0 +1,60 @@
+        var uri = new Uri($"{storage.BaseUri}b/{bucketName}/o/{objectName}?alt=media");
+
+        var request = new HttpRequestMessage { RequestUri = uri };
+        request.Headers.Range = new System.Net.Http.Headers.RangeHeaderValue(firstByte, lastByte);
```
```suggestion
        request.Headers.Range = new RangeHeaderValue(firstByte, lastByte);
```
... and add using...

amanda-tarafa commented on storage/api/Storage.Samples/DownloadByteRange.cs:
```diff
@@ -0,0 +1,60 @@
+        {
+            // Use the HttpClient in the storage object because it supplies
+            // all the authentication headers we need.
+            var response = storage.HttpClient.SendAsync(request).Result;
+            response.Content.CopyToAsync(outputFile, null).Wait();
```
```suggestion
            var response = await storage.HttpClient.SendAsync(request);
            await response.Content.CopyToAsync(outputFile, null);
```

amanda-tarafa commented on storage/api/Storage.Samples/DownloadByteRange.cs:
```diff
@@ -0,0 +1,60 @@
+    /// <param name="firstByte">The first byte of the object content.</param>
+    /// <param name="lastByte">The last byte of the object content.</param>
+    /// <param name="localPath">Local Path where file will be downloaded.</param>
+    public void DownloadByteRange(
```
```suggestion
    public async Task DownloadByteRangeAsync(
```
... and also update the name of the class, file and test class and file.

amanda-tarafa commented on storage/api/Storage.Samples.Tests/DownloadByteRangeTest.cs:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    [Fact]
+    public void DownloadByteRange()
```
This one will now be async as well and return Task.

amanda-tarafa commented on storage/api/Storage.Samples/DownloadEncryptedFile.cs:
```diff
@@ -0,0 +1,49 @@
+        string localPath = null)
+    {
+        var storage = StorageClient.Create();
+        localPath = localPath ?? Path.GetFileName(objectName);
```
Maybe move this to the test code.

amanda-tarafa commented on storage/api/Storage.Samples/DownloadEncryptedFile.cs:
```diff
@@ -0,0 +1,49 @@
+        string key = "3eFsTXPvqi3BpT2ipFCGirslh1Jgc6ikjoAu2oQ5JcI=",
+        string bucketName = "your-unique-bucket-name",
+        string objectName = "my-file-name",
+        string localPath = null)
```
```suggestion
        string localPath = "local-file-path")
```
And check everywhere in case I don't catch them all.

amanda-tarafa commented on storage/api/Storage.Samples/DownloadEncryptedFile.cs:
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.IO;
+
+public class DownloadEncryptedFileSample
```
This one doesn't seem to have a test.

amanda-tarafa commented on storage/api/Storage.Samples/GetRequesterPaysStatus.cs:
```diff
@@ -0,0 +1,39 @@
+        string bucketName = "your-unique-bucket-name")
+    {
+        var storage = StorageClient.Create();
+        var bucket = storage.GetBucket(bucketName, new GetBucketOptions { UserProject = projectId });
```
Do you need to get full projection here?

amanda-tarafa commented on storage/api/Storage.Samples.Tests/RequesterPaysStatusTest.cs:
```diff
@@ -0,0 +1,46 @@
+using Xunit;
+
+[Collection(nameof(BucketFixture))]
+public class RequesterPaysStatusTest
```
Change the class, method and file name to **Get**RequesterPays...

amanda-tarafa commented on storage/api/Storage.Samples.Tests/RequesterPaysStatusTest.cs:
```diff
@@ -0,0 +1,46 @@
+
+        // Enable request pay.
+        var bucket = enableRequesterPaysSample.EnableRequesterPays(_bucketFixture.ProjectId, _bucketFixture.BucketNameGeneric);
+        Assert.True(bucket.Billing.RequesterPays);
```
Don't need to test this here, it should have its own test.

amanda-tarafa commented on storage/api/Storage.Samples.Tests/RequesterPaysStatusTest.cs:
```diff
@@ -0,0 +1,46 @@
+
+        // Disable request pay.
+        bucket = disableRequesterPaysSample.DisableRequesterPays(_bucketFixture.ProjectId, _bucketFixture.BucketNameGeneric);
+        Assert.False(bucket.Billing.RequesterPays);
```
Also don't need to test here it should have its own test as well.

amanda-tarafa commented:
Also, @laljikanjareeya this is looking great, thanks a lot for doing this. Apart from the issues I raised with @JesseLovelace and @frankyn, I don't see any structure issues with the samples or tests, there are a few things that need fixing and some cleanup and nits but overall it's looking really good. I'm fine if you want to start adressing some of my comments before I'm fully done, but it's also fine if you want to wait.

JesseLovelace commented:
re: your feedback @amanda-tarafa 

> I think we need to use Projection.Full on the samples that print metadata for Buckets and Objects, the projection defaults to noAcl but we are printing the ACL information as part of the metadata which would be empty.

I agree, let's put it in

> In some, but not all, of the Update samples, we use metadatageneration to avoid race conditions. I think we should be consistent and either use it on all update samples or on none and have a specific sample showing the use of metageneration. My personal preference is to use it on none of the samples and then have a sample that exclusively uses and explains this, because this is important enough to guarrant its own sample, and because adding it on other samples might be confusing.

I agree, let's remove it from all samples for this pass and add another sample as a follow up

>The usage of UserProject (project that the request will be billed on) is all over the place (some samples use it, some don't), sometimes in places where it makes no sense (like when enabling requester pays). I propose we use it in none of the samples except the actual samples that show how to bill an operation to the requester (RequesterPaysDownload, etc.)

I agree, let's remove it from everywhere that doesn't absolutely need it

> v4 UrlSigning is now the default, but the samples are still explicitly setting the signer version. And the sample that signs with default values does effectivelt the same as the sample that explicitly sets v4. I propose we simplify these and remove the ones that are doing the same. Also, I suppose we don't want v2 signing samples?

We no longer need the V2 signing samples or the sample where the version is unspecified. The v4 sample is the only one that's being published, for all languages; currently the fact that it's the default is unique to C#. For now, let's change the V4 sample to no longer explicitly set the value, but to have a comment that says "C# uses V4 by default".

JesseLovelace commented on storage/api/Storage.Samples/GenerateV4ReadSignedUrl.cs:
```diff
@@ -20,18 +20,13 @@
-        UrlSigner urlSigner = UrlSigner.FromServiceAccountPath(credential);
-        string url = urlSigner.Sign(bucketName, objectName, TimeSpan.FromHours(1), HttpMethod.Get, SigningVersion.V4);
+        UrlSigner urlSigner = UrlSigner.FromServiceAccountPath(credentialFilePath);
+        string url = urlSigner.Sign(bucketName, objectName, TimeSpan.FromHours(1), HttpMethod.Get);
```
Include a comment that says "V4 is the default signing version"

frankyn commented on storage/api/Storage.Samples/GetMetadata.cs:
```diff
@@ -0,0 +1,57 @@
+        Console.WriteLine("Event-based hold enabled? {0}", eventBasedHold);
+        bool temporaryHold = storageObject.TemporaryHold ?? false;
+        Console.WriteLine("Temporary hold enabled? {0}", temporaryHold);
+        Console.WriteLine($"RetentionExpirationTime\t{storageObject.RetentionExpirationTime}");
```
Can you output:

* metadata
* customTime

laljikanjareeya replied:
Added!!

laljikanjareeya commented:
Gentle ping @frankyn @amanda-tarafa @JesseLovelace

amanda-tarafa commented:
Haven't forgotten @laljikanjareeya, just a few other things on my plate.
I've reviewed roughly over half the samples and added comments that apply to individual samples but also comments that will apply to many samples. I don't know if you have addressed all of those already, but if you haven't, I think there's stuff there that will take you some time. I'll get back to this as soon as possible.

laljikanjareeya commented on storage/api/Storage.Samples/DeleteBucket.cs:
```diff
@@ -0,0 +1,33 @@
+    public void DeleteBucket(string bucketName = "your-unique-bucket-name")
+    {
+        var storage = StorageClient.Create();
+        storage.DeleteBucket(bucketName);
```
It's not long running.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/CreateHmacKeyTest.cs:
```diff
@@ -0,0 +1,49 @@
+        HmacKeyDeleteSample hmacKeyDeleteSample = new HmacKeyDeleteSample();
+
+        // These need to all run as one test so that we can use the created key in every test.
+        _bucketFixture.DeleteAllHmacKeys();
```
Removed This Deleted Hmac Key in specific tests.

laljikanjareeya commented on storage/api/Storage.Samples/AddBucketConditionalIamBinding.cs:
```diff
@@ -0,0 +1,68 @@
+            RequestedPolicyVersion = 3
+        });
+
+        policy.Version = 3;
```
Yes if version is not specified it throws below Exception:
```
Specified policy version (1) must be at least 3 based on the policy's contents. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.
```

frankyn replied:
+1, it is necessary unfortunately.

laljikanjareeya replied:
Added Below comment where it's use:
// Set the policy schema version. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.

laljikanjareeya commented:
> I don't know if you have addressed all of those already.

I have addressed all the comments one is pending about KMS, you want me to add KMS library reference?

amanda-tarafa commented on storage/api/Storage.Samples/EnableDefaultKMSKey.cs:
```diff
@@ -0,0 +1,53 @@
+        string kmsKeyRing = "kms-key-ring",
+        string kmsKeyName = "key-name")
+    {
+        string keyPrefix = $"projects/{projectId}/locations/{keyLocation}";
+        string fullKeyringName = $"{keyPrefix}/keyRings/{kmsKeyRing}";
+        string fullKeyName = $"{fullKeyringName}/cryptoKeys/{kmsKeyName}";
```
OK, instead of adding the KMS dependecy for just the resource names, add a comment at the top of these lines:
// If you use the Google.Cloud.Kms.V1 library, you can construct these names using helper classes provided there.

laljikanjareeya replied:
Done!

frankyn commented on storage/api/Storage.Samples/DisableRequesterPays.cs:
```diff
@@ -0,0 +1,50 @@
+        var storage = StorageClient.Create();
+        var bucket = storage.GetBucket(bucketName, new GetBucketOptions
+        {
+            UserProject = projectId
```
Sample only disables requesterPays in other languages and doesn't perform a request with `UserProject`

frankyn commented on storage/api/Storage.Samples/EnableDefaultKMSKey.cs:
```diff
@@ -0,0 +1,53 @@
+        string kmsKeyRing = "kms-key-ring",
+        string kmsKeyName = "key-name")
+    {
+        string keyPrefix = $"projects/{projectId}/locations/{keyLocation}";
```
Just calling out, https://cloud.google.com/storage/docs/encryption/using-customer-managed-keys documentation does call out requirements for the value, e.g. co-location with bucket location and granted permission to the Robot Service account..

frankyn replied:
Documentation does cover it in the set up instructions prior to samples, but not directly in the same area.

![7KHYYkUZw5AVXtW](https://user-images.githubusercontent.com/1907648/94468413-39f59b00-0179-11eb-994c-266b4e3e131a.png)

I think it's fine to keep documentation around setting up the KMS key light-weight for now.

frankyn commented on storage/api/Storage.Samples/ListFilesWithPrefix.cs:
```diff
@@ -0,0 +1,62 @@
+    /// objects in the given "directory". Without the delimiter, the entire  tree
+    /// under the prefix is returned.
+    /// For example, given these objects:
+    ///   /a/1.txt
```
Remove leading `/` from paths because prefix below is `a/`

frankyn commented on storage/api/Storage.Samples/ViewBucketIamMembers.cs:
```diff
@@ -0,0 +1,39 @@
+    public Policy ViewBucketIamMembers(string bucketName = "your-unique-bucket-name")
+    {
+        var storage = StorageClient.Create();
+        var policy = storage.GetBucketIamPolicy(bucketName);
```
Please use: 

```csharp
           var policy = storage.GetBucketIamPolicy(bucketName, new GetBucketIamPolicyOptions()
            {
                RequestedPolicyVersion = 3
            });
```

Do this as well for other IAM add/remove samples. They should look equivalent to samples published here: https://cloud.google.com/storage/docs/access-control/using-iam-permissions

frankyn reviewed (changes requested):
I have a few minor nits, otherwise LGTM.

frankyn commented on storage/api/Storage.Samples/AddBucketIamMember.cs:
```diff
@@ -0,0 +1,43 @@
+        string member = "serviceAccount:dev@iam.gserviceaccount.com")
+    {
+        var storage = StorageClient.Create();
+        var policy = storage.GetBucketIamPolicy(bucketName);
```
Use version 3 pattern for this sample as well.

frankyn commented on storage/api/Storage.Samples/RemoveBucketIamMember.cs:
```diff
@@ -0,0 +1,48 @@
+        string role = "roles/storage.objectViewer",
+        string member = "serviceAccount:dev@iam.gserviceaccount.com")
+    {
+        var storage = StorageClient.Create();
```
Use version 3 pattern here as well.

frankyn replied:
Need to use: 
```
var policy = storage.GetBucketIamPolicy(bucketName, new GetBucketIamPolicyOptions
        {
            RequestedPolicyVersion = 3
        });
        // Set the policy schema version. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.
        policy.Version = 3;
```

frankyn reviewed (changes requested):
2 more nits. Overall LGTM. Thanks @laljikanjareeya

amanda-tarafa reviewed (changes requested):
Mostly nits. Nothing major.
A couple of uses of linq ForEach need to be changed for foreach, etc.
Happy to approve once these changes are in.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/ActivateHmacKeyTest.cs:
```diff
@@ -0,0 +1,55 @@
+        Assert.Equal("ACTIVE", keyMetadata.State);
+
+        // Deactivate key.
+        deactivateHmacKeySample.DeactivateHmacKey(_bucketFixture.ProjectId, key.Metadata.AccessId);
```
Yes we need it deactivate the key before deleting it, but I just move `Deactivate` and `Delete` Logic in one method `DeleteHmacKey`  in Fixture for simplicity.

amanda-tarafa replied:
Perfect, thanks!

laljikanjareeya commented on storage/api/Storage.Samples.Tests/CreateHmacKeyTest.cs:
```diff
@@ -0,0 +1,46 @@
+        Assert.Equal(key.Metadata.ServiceAccountEmail, serviceAccountEmail);
+
+        // Deactivate key.
+        hmacKeyDeactivateSample.DeactivateHmacKey(_bucketFixture.ProjectId, key.Metadata.AccessId);
```
yes need to deactivate before deleting it.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/DeleteHmacKeyTest.cs:
```diff
@@ -0,0 +1,50 @@
+        var key = createHmacKeySample.CreateHmacKey(_bucketFixture.ProjectId, serviceAccountEmail);
+
+        // Deactivate key.
+        deactivateHmacKeySample.DeactivateHmacKey(_bucketFixture.ProjectId, key.Metadata.AccessId);
```
Move this in Fixture.

laljikanjareeya commented on storage/api/Storage.Samples/DisableRequesterPays.cs:
```diff
@@ -0,0 +1,42 @@
+        bucket.Billing.RequesterPays = false;
+        bucket = storage.UpdateBucket(bucket, new UpdateBucketOptions
+        {
+            UserProject = projectId
```
So in All the RequesterPay Sample if it's enable it is necessary to pass `UserProject` for any operation.
without `UserProject` it gives below Error:
```
Bucket is requester pays bucket but no user project provided. [400]
```

laljikanjareeya commented on storage/api/Storage.Samples.Tests/DisableUniformBucketLevelAccessTest.cs:
```diff
@@ -0,0 +1,49 @@
+        // Create bucket
+        createBucketSample.CreateBucket(_bucketFixture.ProjectId, bucketName);
+        _bucketFixture.SleepAfterBucketCreateDelete();
+        _bucketFixture.TempBucketNames.Add(bucketName);
```
Move this logic to Fixture. 👍

laljikanjareeya commented on storage/api/Storage.Samples.Tests/DownloadByteRangeAsyncTest.cs:
```diff
@@ -0,0 +1,42 @@
+        DownloadByteRangeAsyncSample downloadByteRangeSample = new DownloadByteRangeAsyncSample();
+        uploadFileSample.UploadFile(_bucketFixture.BucketNameGeneric, "Resources/HelloDownloadCompleteByteRange.txt",
+            _bucketFixture.Collect("HelloDownloadCompleteByteRange.txt"));
+        await downloadByteRangeSample.DownloadByteRangeAsync(_bucketFixture.BucketNameGeneric, "HelloDownloadCompleteByteRange.txt", 0, 20, "HelloDownloadCompleteByteRange.txt_0-20");
```
If we define the standard line length than in future I will make sure this is follow the standard.

laljikanjareeya commented on storage/api/Storage.Samples/DownloadFileRequesterPays.cs:
```diff
@@ -0,0 +1,37 @@
+        var storage = StorageClient.Create();
+        using (var outputFile = File.OpenWrite(localPath))
+        {
+            storage.DownloadObject(bucketName, objectName, outputFile);
```
Added UserProject Thanks!!

laljikanjareeya commented on storage/api/Storage.Samples.Tests/DownloadFileRequesterPaysTest.cs:
```diff
@@ -0,0 +1,38 @@
+        var downloadFileRequesterPaysSample = new DownloadFileRequesterPaysSample();
+
+        // download file request pays
+        downloadFileRequesterPaysSample.DownloadFileRequesterPays(_bucketFixture.ProjectId, _bucketFixture.BucketNameGeneric, _bucketFixture.FileName, "HelloDownloadObjectRequesterPays2.txt");
```
same this as above about standard line length.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/EnableUniformBucketLevelAccessTest.cs:
```diff
@@ -0,0 +1,48 @@
+        Assert.True(updatedBucket.IamConfiguration.UniformBucketLevelAccess.Enabled);
+
+        // Disable Uniform bucket level access.
+        disableUniformBucketLevelAccessSample.DisableUniformBucketLevelAccess(bucketName);
```
Removed.

laljikanjareeya commented on storage/api/Storage.Samples/GenerateV4UploadSignedUrl.cs:
```diff
@@ -0,0 +1,53 @@
+        };
+
+        UrlSigner.Options options = UrlSigner.Options.FromDuration(TimeSpan.FromHours(1))
+            .WithSigningVersion(SigningVersion.V4);
```
Done!

laljikanjareeya commented on storage/api/Storage.Samples.Tests/GetHmacKeyTest.cs:
```diff
@@ -0,0 +1,50 @@
+        Assert.Equal(keyMetadata.ServiceAccountEmail, serviceAccountEmail);
+
+        // Deactivate key.
+        deactivateHmacKeySample.DeactivateHmacKey(_bucketFixture.ProjectId, key.Metadata.AccessId);
```
same as above comment.

laljikanjareeya commented on storage/api/Storage.Samples/GetRequesterPaysStatus.cs:
```diff
@@ -0,0 +1,36 @@
+        var storage = StorageClient.Create();
+        var bucket = storage.GetBucket(bucketName, new GetBucketOptions
+        {
+            UserProject = projectId
```
same as about comment about requester pay.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/GetUniformBucketLevelAccessTest.cs:
```diff
@@ -0,0 +1,51 @@
+        // Create bucket
+        createBucketSample.CreateBucket(_bucketFixture.ProjectId, bucketName);
+        _bucketFixture.SleepAfterBucketCreateDelete();
+        _bucketFixture.TempBucketNames.Add(bucketName);
```
Moved to Fixture.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/GetUniformBucketLevelAccessTest.cs:
```diff
@@ -0,0 +1,51 @@
+        Assert.True(uniformBucketLevelAccess.Enabled);
+
+        // Disable Uniform bucket level access.
+        disableUniformBucketLevelAccessSample.DisableUniformBucketLevelAccess(bucketName);
```
Removed!!

laljikanjareeya commented on storage/api/Storage.Samples/ListFiles.cs:
```diff
@@ -0,0 +1,37 @@
+        string bucketName = "your-unique-bucket-name")
+    {
+        var storage = StorageClient.Create();
+        var storageObjects = storage.ListObjects(bucketName, "");
```
It was about prefix but we have separate sample for it removed!!

laljikanjareeya commented on storage/api/Storage.Samples.Tests/ListHmacKeysTest.cs:
```diff
@@ -0,0 +1,51 @@
+        Assert.Contains(keys, key => key.AccessId == key.AccessId);
+
+        // Deactivate key.
+        deactivateHmacKeySample.DeactivateHmacKey(_bucketFixture.ProjectId, key.Metadata.AccessId);
```
same as above comment.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/MoveFileTest.cs:
```diff
@@ -0,0 +1,47 @@
+
+        moveFileSample.MoveFile(_bucketFixture.BucketNameGeneric, "HelloMove.txt", _bucketFixture.BucketNameGeneric, _bucketFixture.Collect("ByeMove.txt"));
+
+        // If we try to clean up "HelloMove.txt", it will fail because it moved.
```
It was about removing from collection as we move that file so it will be no more but I remove `Collect` while uploading.
Just Check the exception about 404.
Thanks!!

laljikanjareeya commented on storage/api/Storage.Samples.Tests/PrintFileAclForUserTest.cs:
```diff
@@ -0,0 +1,52 @@
+        RemoveFileOwnerSample removeFileOwnerSample = new RemoveFileOwnerSample();
+
+        string userEmail = _bucketFixture.ServiceAccountEmail;
+        uploadFileSample.UploadFile(_bucketFixture.BucketNameGeneric, _bucketFixture.FilePath, _bucketFixture.Collect("HelloAddObjectOwner.txt"));
```
Done!

laljikanjareeya commented on storage/api/Storage.Samples/RemoveBucketConditionalIamBinding.cs:
```diff
@@ -0,0 +1,59 @@
+              && binding.Condition != null
+              && binding.Condition.Title == title
+              && binding.Condition.Description == description
+              && binding.Condition.Expression == expression).ToList();
```
Needed to add `ToList()` as we are modify the same Collection in loop.
 Gives Below Error:
```
Collection was modified; enumeration operation may not execute.'
```

laljikanjareeya commented on storage/api/Storage.Samples/RemoveBucketIamMember.cs:
```diff
@@ -0,0 +1,48 @@
+                    policy.Bindings.Remove(response);
+                }
+            }
+        });
```
More simplify and added role condition before iteration but need to add `ToList()`.

amanda-tarafa replied:
OK with ToList, but please modify to using foreach(...) instead of the linq ForEach, foreach is the common way to do this.

laljikanjareeya commented on storage/api/Storage.Samples/SetRetentionPolicy.cs:
```diff
@@ -0,0 +1,49 @@
+    /// sets the bucket's retention policy.
+    /// </summary>
+    /// <param name="bucketName">The name of the bucket.</param>
+    /// <param name="retentionPeriod">The bucket's retention policy. The retention policy enforces a minimum retention
```
it's in seconds Added it in description.

laljikanjareeya commented on storage/api/Storage.Samples/RemoveBucketDefaultOwner.cs:
```diff
@@ -0,0 +1,45 @@
+        if (bucket.Acl == null)
+        {
+            bucket.Acl = new List<BucketAccessControl>();
+        }
```
Not needed Removed. Thanks!!

laljikanjareeya commented:
@amanda-tarafa Thanks For the review I have Addressed All the comments. currently, It's failing because of `KMS region` please take a look.

amanda-tarafa commented:
@laljikanjareeya I think it's failing because of the rate at which buckets can be created/delete/updated. I know you were sleeping on creation and deletion, but maybe we need to sleep on update as well. Please try to do so on the test code and not the sample code itself. Also, check the tests that are failing in both Windows and Linux, since that will give you a hint as to which you need to check.

amanda-tarafa reviewed (commented):
A few extra comments. And what I said earlier about the tests failing because of the rate limit, which also appliest to modifying buckets.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/CreateRegionalBucketTest.cs:
```diff
@@ -0,0 +1,38 @@
+    public void CreateRegionalBucket()
+    {
+        CreateRegionalBucketSample createRegionalBucketSample = new CreateRegionalBucketSample();
+        var bucketName = Guid.NewGuid().ToString();
```
Yes added to `TempBuckeNames` List.

laljikanjareeya commented on storage/api/Storage.Samples/DeactivateHmacKey.cs:
```diff
@@ -0,0 +1,43 @@
+        Console.WriteLine($"State: {updatedMetadata.State}");
+        Console.WriteLine($"Time Created: {updatedMetadata.TimeCreated}");
+        Console.WriteLine($"Time Updated: {updatedMetadata.Updated}");
+        Console.WriteLine($"ETag: {updatedMetadata.ETag}");
```
Removed Extra call in Deactivate Hmac Key test.

laljikanjareeya commented on storage/api/Storage.Samples.Tests/EnableRequesterPaysTest.cs:
```diff
@@ -0,0 +1,49 @@
+        Assert.True(bucket.Billing?.RequesterPays);
+
+        // Disable request pay.
+        disableRequesterPaysSample.DisableRequesterPays(_bucketFixture.ProjectId, bucketName);
```
After Disabling the request pay everything works as normal.

laljikanjareeya commented on storage/api/Storage.Samples/RemoveBucketConditionalIamBinding.cs:
```diff
@@ -0,0 +1,59 @@
+            bindingsToRemove.ForEach(binding =>
+            {
+                policy.Bindings.Remove(binding);
+            });
```
Yes Added `foreach` instead of `.ForEach`

laljikanjareeya commented on storage/api/Storage.Samples.Tests/CreateRegionalBucketTest.cs:
```diff
@@ -0,0 +1,38 @@
+    {
+        CreateRegionalBucketSample createRegionalBucketSample = new CreateRegionalBucketSample();
+        var bucketName = Guid.NewGuid().ToString();
+        var buket = createRegionalBucketSample.CreateRegionalBucket(_bucketFixture.ProjectId, bucketName, _bucketFixture.KmsKeyLocation, StorageClasses.Regional);
```
Added `TestLocation` in Fixture.

frankyn commented on storage/api/Storage.Samples/AddBucketConditionalIamBinding.cs:
```diff
@@ -0,0 +1,65 @@
+        string expression = "resource.name.startsWith(\"projects/_/buckets/bucket-name/objects/prefix-a-\")")
+    {
+        var storage = StorageClient.Create();
+        var policy = storage.GetBucketIamPolicy(bucketName);
```
Need to use:
```
var policy = storage.GetBucketIamPolicy(bucketName, new GetBucketIamPolicyOptions
        {
            RequestedPolicyVersion = 3
        });
        // Set the policy schema version. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.
        policy.Version = 3;
```

frankyn commented on storage/api/Storage.Samples/AddBucketIamMember.cs:
```diff
@@ -0,0 +1,46 @@
+        string member = "serviceAccount:dev@iam.gserviceaccount.com")
+    {
+        var storage = StorageClient.Create();
+        var policy = storage.GetBucketIamPolicy(bucketName, new GetBucketIamPolicyOptions
```
Need to use:
```
var policy = storage.GetBucketIamPolicy(bucketName, new GetBucketIamPolicyOptions
        {
            RequestedPolicyVersion = 3
        });
        // Set the policy schema version. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.
        policy.Version = 3;
```

laljikanjareeya replied:
Done!!

frankyn commented on storage/api/Storage.Samples/ViewBucketIamMembers.cs:
```diff
@@ -27,6 +27,9 @@ public Policy ViewBucketIamMembers(string bucketName = "your-unique-bucket-name"
             RequestedPolicyVersion = 3
         });
+        // Set the policy schema version. For more information, please refer to https://cloud.google.com/iam/docs/policies#versions.
+        policy.Version = 3;
```
My mistake, please don't set `policy.Version = 3;` as it's not required to

amanda-tarafa commented:
@laljikanjareeya This is LGTM except for the failure. I can see now the KMS failure and it's related to what @frankyn called out earlier about the KMS key needing to be in the same location as the bucket.
Take a look here: https://cloud.google.com/storage/docs/encryption/using-customer-managed-keys#prereqs, in particular at one of the notes on step 4.

laljikanjareeya commented:
@amanda-tarafa KMS failure has been fixed, now all the storage tests are passing successfully.
The Issue was we are using simple bucket instead of regional bucket that we have created in Fixture.

amanda-tarafa reviewed (approved):
LGTM! Thanks a lot for all the work here @laljikanjareeya .

Let's just wait for @frankyn to approve and confirm that we can merge.

amanda-tarafa commented:
All green @laljikanjareeya , please do the honors :).
(Squash and merge so we don't get all the little in between commits). Thanks!!

laljikanjareeya commented:
> All green @laljikanjareeya , please do the honors :).
> (Squash and merge so we don't get all the little in between commits). Thanks!!

Squash and merge is disabled. I am not able to merge it can you do it from your end please.

amanda-tarafa commented:
Done! I think it's because @JesseLovelace requested some changes and he hasn't approved. But @frankyn approval is enough in that sense.

frankyn commented:
Forgot to mention this in my approval. @JesseLovelace confirmed verbally on LGTM before going on vacation.

Thank you @amanda-tarafa and @laljikanjareeya for all the heavy lifting!
